### PR TITLE
Support Same-DB Ownership Chaining for object references inside Procedure/Functions

### PR DESCRIPTION
**Pull request by HarshLunagariya**
### Description

This commit implements the ownership chaining support for object
references inside function/procedure. The basic idea of Ownership chaining is that when
one object references another, and both have the same owner, then permissions
are only checked when the top-level object is accessed.

Core Functionality:
 Added support for ownership chaining within the same database for procedures and functions

Key Components Modified:
1. Added ownership chain validation in permission checks
2. Walker function to mark relations and functions inside view definitions

Implementation:

Walker function to mark relations and functions inside view definitions

Walks through the query parse tree to:
1. Mark relations and functions as being inside a view context
2. For relations:
   - Set checkAsUser to the view_owner when it matches the relation's owner,
     enabling permission checking to pass at the executor stage (ownership chaining)
3. For procedures/functions:
   - Store the view_owner in the parentOwnerId field to support
     procedure/function-specific ownership chaining logic
     during permission checks at the executor stage

### Issues Resolved

[List any issues this PR will resolve]

### Test Scenarios Covered ###
* **Use case based -**

* **Boundary conditions -**

* **Arbitrary inputs -**

* **Negative test cases -**

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 1]
```diff
@@ -2983,3 +2984,99 @@ downcase_truncate_split_object_name(char *four_part_object_name, char **server_n
...
+ */
+Oid
+get_rel_owner(Oid relid)
+{
+    HeapTuple   tuple;
+    Oid         owner;
+
+    /* Get relation tuple from pg_class */
+    tuple = SearchSysCache1(RELOID, ObjectIdGetDatum(relid));
+    if (!HeapTupleIsValid(tuple))
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c:3001]: Let us add asserts for input not being InvalidOid. Might be needed in other functions as well.

Also let's evaluate if storing the function and relation owners inside a new field in parse nodes strcutures will help performance.

@HarshLunagariya would also like to know your thoughts on this.

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 2]
```diff
@@ -2983,3 +2984,99 @@ downcase_truncate_split_object_name(char *four_part_object_name, char **server_n
...
+ */
+static Oid
+get_current_func_owner(void)
+{
+
+	if (sql_dialect != SQL_DIALECT_TSQL)
+		return InvalidOid;
+
+	if (pltsql_non_tsql_proc_entry_count > 0 || pltsql_sys_func_entry_count > 0)
+		return InvalidOid;
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: Let use `pltsql_support_tsql_transactions()` instead

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 3]
```diff
@@ -1039,13 +1039,27 @@ pltsql_ExecFuncProc_AclCheck(Oid funcid)
 			 */
 			if (nspname != NULL && !is_shared_schema(nspname) &&
 				!is_schema_from_db(schema_id, get_cur_db_id()))
+			{
 				userid = GetSessionUserId();
+			}
+			else
+			{
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/hooks.c:1046]: This block only has check for `IS_TDS_CLIENT() && !InSecurityRestrictedOperation()`. Let's find out the why there is not dialect check here ? based on the answer we might have to remove dialect checks from all functions being used inside this block.

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 4]
```diff
@@ -1039,13 +1039,27 @@ pltsql_ExecFuncProc_AclCheck(Oid funcid)
...
+				/*
+				 * Ownership Chaining Logic for object references inside function/procedures.
+				 * Only applicable for same-db ownership chaining cases.
+				 */
+				if(IsA(expr, FuncExpr) &&
+				   is_valid_func_ownership_chain(expr, get_func_owner(((FuncExpr *)expr)->funcid)))
+				{
+					return ACLCHECK_OK;
+				}
+			}
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/hooks.c:1061]: I think this we should first check if permission are enough without ownership chaining. If we do not have the required permission, only then enter ownership change code. This will reduce the additional sys cache lookups.

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 5]
```diff
@@ -7919,4 +7922,92 @@ gen_command_grant_revoke_priv_to_role(StringInfo query, const char *rolename,
...
+static void
+mark_outside_view(Query *query)
+{
+
+	if (!IS_TDS_CLIENT() || InSecurityRestrictedOperation())
+		return;
+
+	mark_outside_view_ref_walker((Node *)query, NULL);
+}
+
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: Should all of this code go into pltsql_permissions.c ?

**Review by tanscorpio7 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/pltsql_utils.c (hunk 1 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c:3001]: > Also let's evaluate if storing the function and relation owners inside a new field in parse nodes strcutures will help performance.

For each parsenode, Owner will be fetched once at the time of permission check. So I don't there will be much of a performance gain.

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 6]
```diff
@@ -1039,13 +1039,27 @@ pltsql_ExecFuncProc_AclCheck(Oid funcid)
...
+			else
+			{
+				/*
+				 * Ownership Chaining Logic for object references inside function/procedures.
+				 * Only applicable for same-db ownership chaining cases.
+				 */
+				if(IsA(expr, FuncExpr) &&
+				   is_valid_func_ownership_chain(expr, get_func_owner(((FuncExpr *)expr)->funcid)))
+				{
+					return ACLCHECK_OK;
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c:1059]: free nspname

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 7]
```diff
@@ -2983,3 +2984,99 @@ downcase_truncate_split_object_name(char *four_part_object_name, char **server_n
...
+	* Fetch the top procedure excution state from execution state call stack
+	* and get the owner of that procedure. Top entry in stack will have
+	* fn_oid and fn_owner value set.
+	*/
+	if (!exec_state_call_stack ||
+		!exec_state_call_stack->estate ||
+		!exec_state_call_stack->estate->func)
+		return InvalidOid;
+
+	return exec_state_call_stack->estate->func->fn_oid;
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c:3057]: is this really function onwer? I think we need to update function name

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 8]
```diff
@@ -1112,14 +1126,18 @@ pltsql_ExecutorStart(QueryDesc *queryDesc, int eflags)
...
-						if (OidIsValid(perminfo->checkAsUser))
+						if (OidIsValid(perminfo->checkAsUser) && !isTempNamespace(schema_id))
 						{
 							Oid loginId = get_login_for_user(perminfo->checkAsUser, nspname);
 							if (OidIsValid(loginId))
 								perminfo->checkAsUser = loginId;
 						}
 						else
 							perminfo->checkAsUser = GetSessionUserId();
+						if (is_valid_func_ownership_chain(perminfo, get_rel_owner(relOid)))
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c]: nit: get_rel_owner can be reused

File: contrib/babelfishpg_tsql/src/pltsql_utils.c (hunk 1 above)
Comment (tanscorpio7) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c:3001]: Got it.
@pranavJ23 Lets validate the for SELECT f() from table, this permission check code path is only called once and not for each row.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 3 above)
Comment (pranavJ23) [reply contrib/babelfishpg_tsql/src/hooks.c:1046]: We need to have a dialect check where ever we are trying to (re)parse, execute a query in t-sql syntax, since we are checking the user permission when we are connected with bbf client(i.e tds_client) so the IS_TDS_CLIENT() check is sufficient since regardless of dialect we need to do permission check.
Yes we should remove the dialect check from functions called from here since we don't need tsql check(for performing parsing/ executing any tsql query).

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 4 above)
Comment (pranavJ23) [reply contrib/babelfishpg_tsql/src/hooks.c:1061]: right, will add it

File: contrib/babelfishpg_tsql/src/pltsql_utils.c (hunk 7 above)
Comment (pranavJ23) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c:3057]: addressed

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 6 above)
Comment (pranavJ23) [reply contrib/babelfishpg_tsql/src/hooks.c:1059]: addressed

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 9]
```diff
@@ -1039,13 +1039,32 @@ pltsql_ExecFuncProc_AclCheck(Oid funcid)
...
 			if (nspname != NULL && !is_shared_schema(nspname) &&
 				!is_schema_from_db(schema_id, get_cur_db_id()))
+			{
 				userid = GetSessionUserId();
+			}
+			else
+			{
+				/* If user already has the permission then return */
+				if (object_aclcheck(ProcedureRelationId, funcid, userid, ACL_EXECUTE) == ACLCHECK_OK)
+					return ACLCHECK_OK;
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c:1049]: free nspname

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 10]
```diff
@@ -2983,3 +2984,101 @@ downcase_truncate_split_object_name(char *four_part_object_name, char **server_n
...
+}
+
+extern bool
+is_valid_func_ownership_chain(void *expr, Oid objectOwnerId)
+{
+	Oid top_func = InvalidOid;
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: I think we should avoid top function, it may confuse readers. For example, f1 -> f2 -> f3 then f1 might be confused as top function while checking f3. Somthing like immediate_parent_func would make more sense

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 11]
```diff
@@ -1039,13 +1039,32 @@ pltsql_ExecFuncProc_AclCheck(Oid funcid)
...
+				/*
+				 * Ownership Chaining Logic for object references inside function/procedures.
+				 * Only applicable for same-db ownership chaining cases.
+				 */
+				if (IsA(expr, FuncExpr) &&
+				   is_valid_func_ownership_chain(expr, get_func_owner(((FuncExpr *)expr)->funcid)))
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c:1055]: IIUC, there will be two cases here
1. Function which are marked inside view. 
2. Function which are marked outside view 

For function marked outside of view, I can understand how will it behave. But what about function marked inside view and has a parent owner set? How are we checking permission in that case?

**Review by Deepesh125 — CHANGES_REQUESTED** [review]

**Review by Deepesh125 — DISMISSED** [review]

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 11 above)
Comment (tanscorpio7) [reply contrib/babelfishpg_tsql/src/hooks.c:1055]: @pranavJ23 Let us push `get_func_owner` to inside `is_valid_func_ownership_chain`. In many cases `is_valid_func_ownership_chain` will exit even before requiring the owner info. We should only do sys cache lookup for owner when we really need it.

**Review by tanscorpio7 — CHANGES_REQUESTED** [review]

**Review by tanscorpio7 — APPROVED** [review]

**Review by Deepesh125 — APPROVED** [review]
